lib/cppcheck.cpp

```diff
@@ -754,6 +754,8 @@ unsigned int CppCheck::checkFile(const FileWithDetails& file, const std::string
             toolinfo << (mSettings.severity.isEnabled(Severity::portability) ? 'p' : ' ');
             toolinfo << (mSettings.severity.isEnabled(Severity::information) ? 'i' : ' ');
             toolinfo << mSettings.userDefines;
+            toolinfo << std::to_string(static_cast<std::uint8_t>(mSettings.checkLevel));
+            // TODO: do we need to add more options?
```

Comment:
I filed https://trac.cppcheck.net/ticket/13379 about this.

Reply:
That would be a rather pointless test. It returns exactly what we expect it to return based on the input so it would just be a logic duplication.

There's more refactorings coming so maybe one of those will give us an in for a useful test.

Reply:
your test only checks that --check-level=normal / --check-level=exhaustive works as expected. With such unit test we can more quickly test that all the flags we want to test influence the output. I would not think that pytest tests that cover each flag would be nice. 1 pytest that checks that the everything works as expected if toolinfo is different is enough imho.

Reply:
But the test is rather pointless as you need to replicate the whole logic of the function. And it would not be the final hash. Also that test would break soon as the suppressions need to be split from the settings and would just complicate things. 

And we should actually have not a single test in Python but one for each possible values as each of them might have side effects which would not be visible if you try to test it as whole.

This is only addressing part of this because a user encountered this. This needs a lot more work and a first step which tests this in a more generic will follow soon.

Reply:
> But the test is rather pointless as you need to replicate the whole logic of the function. 

So how about something like:
```
    std::set<std::string> toolinfo;
    ASSERT(toolinfo.insert(SettingsBuilder().build().toolinfo());
    ASSERT(toolinfo.insert(SettingsBuilder().checkLevel(exhaustive).build().toolinfo());
    ASSERT(toolinfo.insert(SettingsBuilder().enable(warning).build().toolinfo());
    ASSERT(toolinfo.insert(SettingsBuilder().enable(style).build().toolinfo());
    ASSERT(toolinfo.insert(SettingsBuilder().enable(information).build().toolinfo());
    ... etc ...
```
I also feel that we shouldn't replicate the whole logic of the function.

Reply:
That also just duplicates the logic. And I do not want to expose `toolinfo()` as it is something arbitrary. It should rather expose the hash. That logic should live in `AnalyzerInformation` anyways.

But as I said any such changes will cause problems when `Suppressions` is split from `Settings`.

CppCheck has too much logic anyways which it does not need but that needs several refactorings before that can finally be addressed.

Reply:
I still rather have unit test(s) to test that different settings are covered than having pytest tests for that purpose. I just want 1 generic pytest to test that it all works. Don't we already have a pytest that test that it works properly if for instance the severities are changed?

> That also just duplicates the logic

then I don't understand what you mean. There is no Settings::toolinfo() implementation in the test, I thought you were talking about replicating the implementation.

> And I do not want to expose toolinfo() as it is something arbitrary. It should rather expose the hash.

The hash depends on other things. The Cppcheck version, the suppressions. The hash is even more arbitrary. 

We can implement `calculateHash` like so:
```
static std::size_t calculateHash(const Preprocessor& preprocessor, const simplecpp::TokenList& tokens, const Settings& settings)
{
    std::ostringstream toolinfo;
    toolinfo << CPPCHECK_VERSION_STRING;
    toolinfo << settings.toolinfo();
    settings.supprs.nomsg.dump(toolinfo);
    return preprocessor.calculateHash(tokens, toolinfo.str());
}
```
It separates the concerns as this calculation in analyzerinfo doesn't have to be modified if we add some more Settings flag.

Reply:
btw.. I don't know if I like the name `toolinfo` if we expose it as a method. But I don't see a better name directly.

Reply:
If you are really against the Settings::toolinfo().. if you expose a `calculateHash` in analyzerinfo and unit test that then I am also happy. I want to have unit tests that is my motivation.

Reply:
That requires some refactoring first.

I would like to merge this fix before that as it was reported by a user and not discovered internally.

Reply:
I am not sure the testing in other_test.py should be added. Probably there is a similar test already.

Reply:
okay.. let's refactor after merge..

Reply:
We can still remove/merge it later on - more things need to be tested in future anyways. Too much tests is usually not the issue we are having. 🤪

Reply:
> Too much tests is usually not the issue we are having.

Yes I do agree however I want to also say I would not like to just throw in lots of random tests in random places. Let's continue to structure it as much as we can and make the tests as clean and simple as possible.

As far as I see there is no existing pytest test for calculateHash yes that is really unfortunate.